zcash_client_backend/src/data_api/chain.rs

```diff
@@ -355,6 +394,15 @@ where
             )
             .map_err(Error::Scan)?;
 
+            let (s, r) = scanned_block
+                .transactions
+                .iter()
+                .fold((0, 0), |(s, r), wtx| {
+                    (s + wtx.sapling_spends.len(), r + wtx.sapling_outputs.len())
+                });
+            spent_note_count += s;
```

Comment:
This undercounts during reverse scanning, so it isn't a reliable metric. We either need to exclude it, document it as such, or additionally modify `put_blocks` to report how many spends it detects for the reverse case (so that this metric overall reports all spends, just not necessarily while scanning the block ranges that actually perform the spends).

Reply:
What we're trying to determine here is the answer to the question "do I need to update my user's view of the wallet state". We could just return a boolean, but that seems a bit impoverished. When we're scanning in reverse order and discover one of our notes for which the nullifier has already been revealed in a spend in a later block, we're actually discovering a received note, and we always want to update the user's view of the wallet state when we find a received not, irrespective of whether that note has been spent. So, I see two possibilities. We can keep this as it is implemented here, or we can modify `put_blocks` as you suggest, counting the received notes in two separate buckets, one containing those that we know have already been spent, and the other where we're not sure whether they've already been spent.

Since the "received, but we're not sure if it's been spent" bucket would have uncertainty more or less equivalent to the undercounting of spends you describe in your comment above, I think that we should just keep this implementation.

Reply:
I've updated the documentation of `ScanSummary` to clarify the meaning of these values.

Reply:
The problem here is that because `ScanSummary` is completely undocumented, I have no way of knowing what `ScanSummary::spent_sapling_note_count` is meant to mean, beyond what the method name says. And what the method name says, is not what it actually counts.

If these numbers are intended to be "something happened, and here is a measure of how much of that something happened in this range", then we should convey that in the output (similarly to how the progress metric is an opaque numerator and denominator, and does not indicate to the caller that it happens to be counting notes).

Reply:
Changes post-documentation LGTM.